// raft/raft.proto
syntax = "proto3";

package raft;

// Single log entry.
message LogEntry {
  int32 index = 1;      // k
  int32 term = 2;       // t
  string operation = 3; // o
}

// RequestVote RPC ====
message RequestVoteRequest {
  string candidate_id = 1;
  int32 term = 2;
}

message RequestVoteResponse {
  bool vote_granted = 1;
  int32 term = 2;
}

// AppendEntries RPC (also used as heartbeat) ====
message AppendEntriesRequest {
  string leader_id = 1;
  int32 term = 2;
  repeated LogEntry entries = 3; // full log for simplicity
  int32 commit_index = 4;        // c
}

message AppendEntriesResponse {
  bool success = 1;
  int32 term = 2;
}

// Client operation (e.g., "SET x", "ADD_SECRET", etc.)
message ClientOperationRequest {
  string client_id = 1;
  string node_id = 2;     // node that receives the client request
  string operation = 3;
}

message ClientOperationResponse {
  bool success = 1;
  string result = 2;
  string leader_id = 3;   // who is leader (for redirection)
}

// Main Raft service.
service RaftService {
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc ClientRequest(ClientOperationRequest) returns (ClientOperationResponse);
}
